Initial commit to add phase adjustment feature

mattieudv opened the pull request:
Please have a look through the changes I made. I think it is everything that is needed except for stuff related to parmfiles.

mattieudv committed:
Initial commit to add phase adjustment feature

mauch commented on scripts/KATCalPipe.py:
```diff
@@ -32,14 +32,15 @@
 parser.add_option("--katdal_timeout", default=300, type='int', help='Timeout (in seconds) for data read in katdal')
 parser.add_option("--katdal_retries", default=2, type='int', help='Number of retries for data read in katdal')
+parser.add_option("--antphase-adjust-filename", default=None, help="string or None\nfilename to numpy .npz file containing phase adjustment per input per channel\nto be applied to all raw visibilities before further processing.")
 (options, katfilenames) = parser.parse_args()
 
 if len(katfilenames) == 0:
     parser.print_help()
     sys.exit()
 
 kwargs = {}
-for k in ['parmFile', 'scratchdir', 'targets', 'configFile', 'timeav', 'flag', 'reuse', 'zapraw', 'aipsdisk', 'halfstokes', 'gzip', 'dropants', 'blmask', 'refant', 'katdal_refant', 'polcal', 'XYtarg', 'delaycal_mvf', 'katdal_timeout', 'katdal_retries']:
+for k in ['parmFile', 'scratchdir', 'targets', 'configFile', 'timeav', 'flag', 'reuse', 'zapraw', 'aipsdisk', 'halfstokes', 'gzip', 'dropants', 'blmask', 'refant', 'katdal_refant', 'polcal', 'XYtarg', 'delaycal_mvf', 'katdal_timeout', 'katdal_retries', 'antphase_adjust_filename']:
```
I think we should add it as a parameter file argument rather than a command line argument.

You just need to add a new parameter to the dictionary defined in https://github.com/mauch/KATObitPipe/blob/master/katim/KATCal.py#L33 

It will be picked up and the default overridden when someone supplies a parameter file.

mauch commented on katim/KATH5toAIPS.py:
```diff
@@ -601,12 +603,33 @@ def ConvertKATData(outUV, katdata, meta, err, static=None, blmask=1.e10, stop_w=
-        OErr.PLog(err, OErr.Info, msg)
-        OErr.printErr(err)
-        print(msg)
+    if antphase_adjust_filename is not None:
+        #dictionary with items e.g. antphasedict_rad={'freqMHz':np.linspace(856,1712,4096,endpoint=False),'m000h':np.zeros(nchans),'m000v':np.zeros(nchans)}
+        antphasedict_rad=dict(np.load(antphase_adjust_filename))
+        #check all is OK (any inputs missing, channels match)
+        missinginputs=[]
+        for input0 in h5.inputs:
+            if input0 not in antphasedict_rad:
+                missinginputs.append(input0)
+        if len(missinginputs):
+            msg='Warning: The phase adjustment is set to zero for the following inputs that are missing from %s: %s'%(antphase_adjust_filename,','.join(missinginputs))
+            OErr.PLog(err, OErr.Info, msg)
+            OErr.printErr(err)
+            print(msg)
+            for input0 in missinginputs:
+                antphasedict_rad[input0]=np.zeros(len(antphasedict_rad['freqMHz']))
+        if (h5.channel_width/1e6!=antphasedict_rad['freqMHz'][1]-antphasedict_rad['freqMHz'][0]):
+            msg='Error: channel_width (%g MHz) does not match that of %s (%g MHz)'%(h5.channel_width/1e6,antphase_adjust_filename,antphasedict_rad['freqMHz'][1]-antphasedict_rad['freqMHz'][0])
+            OErr.PLog(err, OErr.Info, msg)
+            OErr.printErr(err)
+            print(msg)
+        if (h5.channels[-1]>len(antphasedict_rad['freqMHz'])):
+            msg='Error: incorrect number of channels (%d) in %s'%(len(antphasedict_rad['freqMHz']),antphase_adjust_filename)
+            OErr.PLog(err, OErr.Info, msg)
+            OErr.printErr(err)
+            print(msg)
+    else:
+        antphasedict_rad=None
```
This probably should be in its own function.

mauch commented on katim/KATH5toAIPS.py:
```diff
@@ -646,6 +669,9 @@ def ConvertKATData(outUV, katdata, meta, err, static=None, blmask=1.e10, stop_w=
             if doweight==False:
                 wt[:] = 1.
             vs = scan_vs[:nint]
+            if antphasedict_rad is not None:
+                for iprod,(input0,input1) in enumerate(katdata.corr_products):
+                    vs[:,:,iprod]*=np.exp(1j*(antphasedict_rad[input0][katdata.channels]-antphasedict_rad[input1][katdata.channels]))[np.newaxis,:]
```
Maybe I'm missing something, but couldn't you have the function above that generates error message also generate the multiplier (` np.exp(1j*(antphasedict_rad[input0][katdata.channels]-antphasedict_rad[input1][katdata.channels]))[np.newaxis,:]` ) so that it can broadcast into vs. This would speed up the calculation a lot.

mauch replied:
Actually, thinking about it some more maybe not since `nchan,ncorr` might be too much memory in general.

mauch reviewed (commented):
Let chat about parameter files when you get the chance :)

mattieudv committed:
addressing PR comments

mattieudv committed:
addressing PR2 comments

Commit:
Small changes to get it working.

1. Actually pass the required parm to KATH5toAIPS
2. Pass Obit logger to new function
3. Change the structure of load_antphasedict so that it stays on the 'happy path' and actually populates antphasedict_exp.

Commit:
Fix a numpy misnaming.

mauch commented:
@mattieudv I've made some extra changes so that it will actually run and some minor other things (see the commit message). Can you let me know if you are OK with the changes I have made?
I'm running it now - should take an hour or so. I'll let you know how it goes.

mattieudv commented:
Looks good.

Commit:
Pre-compute the phase correction over the corr products.

Only adds <500MB of ram and speeds up the processing by more than a factor of 10.

Commit:
Further changes based on discussion with Mattieu:

1) Add a missing 1j to the exponent in the phase adjustment calculation.
2) Expose the quack as a parameter (so the user can select the number of
   dumps to reject from the beginning of each scan)
3) Initialize the static flags across baselines (like I did for the phase adjustment).
   This also uses little ram, and makes the static flag application 20x faster.

mauch commented:
@mattieudv I've pushed some more changes. Can you check these ones. And if you are happy I'll approve the PR.

BTW - Dont worry too much about the changes to the static flags stuff - I've just realized that I can do a similar thing there to what I did with the phase adjustment correction (i.e. initialise it with channel and baseline and only broadcast the time axis) to get a significant speed up in the UV conversion routine.

mattieudv commented:
I think Bill uses the quack parameters, so you would have to warn him that it is changed from boolean to 1 in his scripts going forward. He might not like this.

mauch commented:
That parameter was completely unused before (I had removed the entire 'Quack' functionality from the pipeline years ago and hardcoded a quack=1 at the katdal read step.
Now I've chaged that paremeter so that it does something again - i.e. defines as an int the number of dumps to quack and remove that many from the katdal read step.

mattieudv commented:
Looks good
